File: vesta_si_erpnext/vesta_si_erpnext/report/inspection_control_chart_based_on_permissible_limits/inspection_control_chart_based_on_permissible_limits.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
	columns, data = [], []
	data = get_data(filters)
	columns = get_columns(filters)
	chart = prepare_chart(data)
	return columns, data, None, chart

def get_data(filters):
	condition = ''
	if filters.get('item_code'):
		condition += f" and qi.item_code = '{filters.get('item_code')}'"
	if filters.get('from_date'):
		condition += f" and qi.report_date >= '{filters.get('from_date')}'"
	if filters.get('to_date'):
		condition += f" and qi.report_date <= '{filters.get('to_date')}'"
	if filters.get('qi_parameter'):
		condition += f" and qri.specification = '{filters.get('qi_parameter')}'"
	
	qi_data = frappe.db.sql(f"""Select qi.name, qi.item_code, qri.specification, qri.min_value, qri.max_value, qri.reading_1
								From `tabQuality Inspection` as qi
								Left Join `tabQuality Inspection Reading` as qri ON qri.parent = qi.name
								Where qi.docstatus = 1 and qri.numeric = 1 {condition}
							""",as_dict =1)
	
	return qi_data

def get_columns(filters):
	columns = [
		{
			"fieldname":"name",
			"label": _("Quality Inspection"),
			"fieldtype": "Link",
			"options": "Quality Inspection",
			"width":150
		},
		{
			"fieldname":"item_code",
			"label": _("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			"width":150
		},
		{
			"fieldname":"specification",
			"label": _("QI Parameter"),
			"fieldtype": "Link",
			"options": "Quality Inspection Parameter",
			"width":150
		},
		{
			"fieldname":"reading_1",
			"label": _("Reading"),
			"fieldtype": "Float",
			"width":150
		},
		{
			"fieldname":"max_value",
			"label": _("Maximum Value"),
			"fieldtype": "Float",
			"width":150
		},
		{
			"fieldname":"min_value",
			"label": _("Minimum Value"),
			"fieldtype": "Float",
			"width":150
		},
	]
	
	return columns

def prepare_chart(data):
	return {
			
			"data": {
					'labels': [row.name for row in data],
					'datasets': [
						{
							'name': 'Max Value',
							'values': [row.max_value for row in data],
							'chartType': 'line',
						},
						{
							'name': "Reading",
							'values': [row.reading_1 for row in data],
							'chartType': 'line',
						},
						{
							'name': "Min Value",
							'values': [row.min_value for row in data],
							'chartType': 'line',
						},
						
					]
				},
				"type": "line",
				"colors": ["#008000", "#0000FF", "#FF0000"],
			}

File: vesta_si_erpnext/vesta_si_erpnext/report/inspection_control_chart_based_on_permissible_limits/inspection_control_chart_based_on_permissible_limits.js
// Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Inspection Control Chart Based On Permissible Limits"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname":"group_by",
			"label": __("Group By"),
			"fieldtype": "Select",
			"default":"Item Code",
			"options": ["Item Code", "Item Group"],
			on_change: function () {
				frappe.query_report.toggle_filter_display(
					"item_code",
					frappe.query_report.get_filter_value("group_by") == "Item Group"
				);
				frappe.query_report.toggle_filter_display(
					"item_group",
					frappe.query_report.get_filter_value("group_by") == "Item Code"
				);
				frappe.query_report.set_filter_value("item_group", "");
				frappe.query_report.set_filter_value("item_code", "");
			},
			onload: function () {
				frappe.query_report.toggle_filter_display(
					"item_code",
					frappe.query_report.get_filter_value("group_by") == "Item Group"
				);
				frappe.query_report.toggle_filter_display(
					"item_group",
					frappe.query_report.get_filter_value("group_by") == "Item Code"
				);
				frappe.query_report.set_filter_value("item_group", "");
				frappe.query_report.set_filter_value("item_code", "");
			},
		},
		{
			"fieldname":"item_code",
			"label": __("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			"get_query": function() {
				return {
					filters: {
						"has_batch_no": 1
					}
				};
			},
		},
		{
			"fieldname":"item_group",
			"label": __("Item Group"),
			"fieldtype": "Link",
			"options": "Item Group",
		},
		{
			"fieldname":"qi_parameter",
			"label": __("QI Parameter"),
			"fieldtype": "Link",
			"options": "Quality Inspection Parameter",
		},
		{
			"fieldname":"workstation",
			"label":__("Workstation"),
			"fieldtype":"Link",
			"options":"Workstation"
		}
	],
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (data.max_value < data.reading_1){
			value = value.replace('<a ', '<a style="color: red;" ')
			value = "<span style='color:red'>" + value + "</span>";
		}
		if(data.min_value > data.reading_1){
			value = value.replace('<a ', '<a style="color: red;" ')
			value = "<span style='color:red'>" + value + "</span>";
		}
			
		return value
	}
};
